sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/RowSerializerPool.scala

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.spark
+
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.catalyst.encoders.{ExpressionEncoder, RowEncoder}
+import org.apache.spark.sql.types.StructType
+
+import java.time.Instant
+import java.util.concurrent.{ConcurrentLinkedQueue, Executors, TimeUnit}
+import java.util.concurrent.atomic.AtomicLong
+import scala.collection.concurrent.TrieMap
+
+/**
+ * Spark serializers are not thread-safe - and expensive to create (dynamic code generation)
+ * So we will use this object pool to allow reusing serializers based on the targeted schema.
+ * The main purpose for pooling serializers (vs. creating new ones in each PartitionReader) is for Structured
+ * Streaming scenarios where PartitionReaders for the same schema could be created every couple of 100
+ * milliseconds
+ * A clean-up task is used to purge serializers for schemas which weren't used anymore
+ * For each schema we have an object pool that will use a soft-limit to limit the memory footprint
+ */
+private object RowSerializerPool {
+  val MaxPooledSerializerCount = 256
+  private[this] val cleanUpIntervalInSeconds = 300
+  private[this] val expirationIntervalInSeconds = 1800
+  private[this] val schemaScopedSerializerMap =
+    new TrieMap[StructType, RowSerializerQueue]
+  private[this] val executorService = Executors.newSingleThreadScheduledExecutor(new DaemonThreadFactory())
+
+  executorService.scheduleWithFixedDelay(
+    () => this.onCleanUp(),
+    cleanUpIntervalInSeconds,
+    cleanUpIntervalInSeconds,
+    TimeUnit.SECONDS)
+
+  def getOrCreateSerializer(schema: StructType): ExpressionEncoder.Serializer[Row] = {
+    schemaScopedSerializerMap.get(schema) match {
+      case Some(objectPool) => objectPool.borrowSerializer(schema)
+      case None => RowEncoder(schema).createSerializer()
+    }
+  }
+
+  def returnSerializerToPool(schema: StructType, serializer: ExpressionEncoder.Serializer[Row]): Boolean = {
+    schemaScopedSerializerMap.get(schema) match {
+      case Some(objectPool) => objectPool.returnSerializer(serializer)
+      case None =>
+        val newQueue = new RowSerializerQueue()
+        newQueue.returnSerializer(serializer)
+        !schemaScopedSerializerMap.putIfAbsent(schema, newQueue).isDefined
```

Comment:
Should the put be happening on the borrow instead?

Reply:
I don't think functionally it makes a difference - motivation for putting it here was that I think the risk of race conditions (where putIfAbsence would find a concurrently added serializer) is lower when doing it on task completion than at the beginning of a job. That race condition shouldn't cause any issues - so I don't have a strong preference thought. Why owuld you prefer it on borrow?

Reply:
Well, it's the pattern I've seen most used, if the creation of the resource is what is expensive, then gate the creation of the resource and avoid extra creations. That way, in a concurrent scenario of multiple partition readers start, you avoid the cost of creation.

Reply:
Resolved offline - The issue is the serializer is not thread-safe.

Reply:
Discussed offline with Matias - we agree leaving it here is fine. Main misunderstand was that the Serializer isn't thread-safe and so we cannot allow it to be used across PartitionReaders. And also the cost of creating a Serializer was definitely too high when doing it for every Row - but having a few race conditions where we might create a couple more Serializer instances than strictly required isn't the end of the world. The object pooling will help over time still significantly.